drivers: sensor: lsm6dso: add support for tilt and tap triggers

ivanwagner opened the pull request:
This patch adds support for tilt, tap and double tap triggers. As an example ST IIS2DLPC sensor was taken as an example. In fact DT new properties are taken from there with the same philosophy.

A new sensor trigger value has been added: SENSOR_TRIG_TILT

When the interrupt handler is called, access to 5 registers will determine information regarding any interrupt flag available on this devices. This adds extensibility for future features at a cost of accessing 5 registers.

ivanwagner committed:
drivers: sensor: lsm6dso: add support for tilt and tap triggers

Tilt or Tap can be selected via config and further tap parameters can be
configured via DT.

Signed-off-by: Ivan Wagner <ivan.wagner@tecinvent.ch>

MaureenHelm reviewed (dismissed):
@avisconti can you take a look?

kartben commented on drivers/sensor/st/lsm6dso/lsm6dso_trigger.c:
```diff
@@ -48,6 +48,193 @@ static int lsm6dso_enable_t_int(const struct device *dev, int enable)
+	int ret = 0;
+	lsm6dso_emb_sens_t sens;
+
+	sens.tilt = PROPERTY_ENABLE;
```
should this be as below? this way you save power when turning off interrupt by disabling more than just the routing?

```suggestion
	sens.tilt = enable;
```

avisconti commented on drivers/sensor/st/lsm6dso/lsm6dso_trigger.c:
```diff
@@ -175,36 +412,62 @@ static void lsm6dso_handle_interrupt(const struct device *dev)
-		if (lsm6dso_status_reg_get(ctx, &status) < 0) {
-			LOG_DBG("failed reading status reg");
+
+		if (lsm6dso_all_sources_get(ctx, &sources) < 0) {
```
THis routine is really expensive in terms of bus cycles. I think that maybe we should avoid calling it if TAP and TILT are not enabled... What do you think?

ivanwagner replied:
Yes I agree. I'll make some tests. Unfortunately tap and tilt are not available via lsm6dso_status_reg_get (it's how ST's driver is made). I will split it so that if tap and tilt are not enabled it will run the existing code accessing lsm6dso_status_reg_get.

ivanwagner replied:
I split the code in two distinct handlers depending if tilt or tap are used. It seemed tricky to me to have a single function combining the two. Let me know what you think.

ivanwagner replied:
Good to know. Thanks.

kartben commented:
@ivanwagner can you take a look at the recent comments? Thanks!

ivanwagner commented on drivers/sensor/st/lsm6dso/lsm6dso_trigger.c:
```diff
@@ -48,6 +48,193 @@ static int lsm6dso_enable_t_int(const struct device *dev, int enable)
+
+	}
+
+	return 0;
```
I don't agree with this one. It's reachable when everything does fine.

ivanwagner committed:
drivers: sensor: lsm6dso: improvements on tilt and tap triggers

Improvements include: check error codes, log level adjustments,
removed unreachable code, disable tilt when not used, call optimized
interrupt handler when tilt and tap are not used.

Signed-off-by: Ivan Wagner <ivan.wagner@tecinvent.ch>

yperess commented on drivers/sensor/st/lsm6dso/lsm6dso_trigger.c:
```diff
@@ -206,10 +516,11 @@ static void lsm6dso_handle_interrupt(const struct device *dev)
 
-	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy,
-					GPIO_INT_EDGE_TO_ACTIVE);
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy, GPIO_INT_EDGE_TO_ACTIVE);
```
Generally you don't want to turn on/off the edge triggers. This is outside of the scope of this PR, but I recommend revisiting this for this driver. On level interrupts you want to turn it off because it'll keep firing until you read the status register. With edge interrupts it's safe to leave it on this way you're not missing events.

avisconti commented on drivers/sensor/st/lsm6dso/lsm6dso_trigger.c:
```diff
@@ -206,10 +516,11 @@ static void lsm6dso_handle_interrupt(const struct device *dev)
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_drdy, GPIO_INT_EDGE_TO_ACTIVE);
 }
 
+#endif
```
Put
```
#endif /* defined(CONFIG_LSM6DSO_TILT) || defined(CONFIG_LSM6DSO_TAP) */
```

avisconti reviewed (changes requested):
Look almost good. Just small nits

ivanwagner committed:
drivers: sensor: lsm6dso: added comments in preproc endif

Added comments ad the end of preprocessor if else endif blocks

Signed-off-by: Ivan Wagner <ivan.wagner@tecinvent.ch>
